src/hotspot/share/opto/reachability.cpp

```diff
@@ -418,72 +301,109 @@ static void enumerate_interfering_sfpts(ReachabilityFenceNode* rf, PhaseIdealLoo
 
     if (idx < cur->req()) {
       stack.set_index(idx + 1);
-      linear_traversal(cur->in(idx), stack, visited, safepoints);
+      enumerate_interfering_sfpts_linear_traversal(cur->in(idx), stack, visited, interfering_sfpts);
     } else {
       stack.pop();
     }
   }
+  // Reset temporary structures to their initial state.
+  assert(stack.is_empty(), "required");
+  visited.clear();
 }
 
 // Start offset for reachability info on a safepoint node.
 static uint rf_base_offset(SafePointNode* sfpt) {
   return sfpt->jvms()->oopoff();
 }
 
+static bool dominates_another_rf(ReachabilityFenceNode* rf, PhaseIdealLoop* phase) {
+  assert(!rf->is_redundant(phase->igvn()), "");
+
+  for (int i = 0; i < phase->C->reachability_fences_count(); i++) {
+    ReachabilityFenceNode* other_rf = phase->C->reachability_fence(i);
+    assert(other_rf->outcnt() > 0, "dead node");
+    if (rf != other_rf && rf->referent()->eqv_uncast(other_rf->referent()) &&
+        phase->is_dominator(rf, other_rf)) {
+      return true; // dominates another reachability fence with the same referent
+    }
+  }
+  return false;
+}
+
 // Phase 2: migrate reachability info to safepoints.
 // All RFs are replaced with edges from corresponding referents to interfering safepoints.
 // Interfering safepoints are safepoint nodes which are reachable from the RF to its referent through CFG.
```

Comment:
Seems you don't do it for ALL any more, you drop those that `dominates_another_rf`. You should probably adapt the comment here.